service/src/main/java/bio/terra/catalog/datarepo/DatarepoService.java

```diff
@@ -76,18 +90,21 @@ public SnapshotPreviewModel getPreviewTable(
     }
   }
 
-  private List<String> rolesForAction(SamAction action) {
-    return switch (action) {
-      case READ_ANY_METADATA -> READER_ROLES;
-      case CREATE_METADATA, DELETE_ANY_METADATA, UPDATE_ANY_METADATA -> OWNER_ROLES;
-    };
-  }
-
-  public boolean userHasAction(AuthenticatedUserRequest user, String snapshotId, SamAction action) {
+  public DatasetAccessLevel getRole(AuthenticatedUserRequest user, String snapshotId) {
     try {
       UUID id = UUID.fromString(snapshotId);
-      var roles = rolesForAction(action);
-      return snapshotsApi(user).retrieveUserSnapshotRoles(id).stream().anyMatch(roles::contains);
+      List<String> roles = snapshotsApi(user).retrieveUserSnapshotRoles(id);
+      for (DatasetAccessLevel datasetAccessLevel : DatasetAccessLevel.values()) {
+        if (roles.stream()
+            .map(SNAPSHOT_ROLE_TO_DATASET_ACCESS_LEVEL::get)
+            .anyMatch(
+                roleAsDatasetAccessLevel -> roleAsDatasetAccessLevel.equals(datasetAccessLevel))) {
+          return datasetAccessLevel;
+        }
```

Comment:
If this was a function, of TDR roles (List<String>) to catalog role (DatasetAccessLevel), I think you could use it in `getSnapshotIdsAndRoles()` above and not duplicate the logic.